// @lwc/engine/lib/src/env/window.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
let MO = window.MutationObserver;
// MutationObserver is not yet implemented in jsdom:
// https://github.com/jsdom/jsdom/issues/639
if (typeof MO === 'undefined') {
    /* tslint:disable-next-line:no-empty */
    function MutationObserverMock() { }
    MutationObserverMock.prototype = {
        observe() {
            if (process.env.NODE_ENV !== 'production') {
                if (process.env.NODE_ENV !== 'test') {
                    throw new Error(`MutationObserver should not be mocked outside of the jest test environment`);
                }
            }
        }
    };
    MO = window.MutationObserver = MutationObserverMock;
}
const MutationObserver = MO;
exports.MutationObserver = MutationObserver;
const MutationObserverObserve = MutationObserver.prototype.observe;
exports.MutationObserverObserve = MutationObserverObserve;
let { addEventListener: windowAddEventListener, removeEventListener: windowRemoveEventListener, } = window;
exports.windowAddEventListener = windowAddEventListener;
exports.windowRemoveEventListener = windowRemoveEventListener;
/**
 * This trick to try to pick up the __lwcOriginal__ out of the intrinsic is to please
 * jsdom, who usually reuse intrinsic between different document.
 */
// @ts-ignore jsdom
exports.windowAddEventListener = windowAddEventListener = windowAddEventListener.__lwcOriginal__ || windowAddEventListener;
// @ts-ignore jsdom
exports.windowRemoveEventListener = windowRemoveEventListener = windowRemoveEventListener.__lwcOriginal__ || windowRemoveEventListener;
